dqc_us_rules/dqc_us_0018.py

```diff
@@ -0,0 +1,329 @@
+                                ('', ''))[1]
+                        )
+                json_str = str(json.dumps(
+                    val.usgaapDeprecations,
+                    ensure_ascii=False,
+                    indent=0)
+                )  # might not be unicode in 2.7
+                saveFile(cntlr, deprecations_json_file, json_str)
+                deprecations_instance.close()
+                del deprecations_instance  # dereference closed modelXbrl
+            val.modelXbrl.profileStat(
+                _("build us-gaap deprecated concepts cache"),  # noqa
+                time.time() - started_at
+            )
+        year += 1
+
+
+def deprecated_facts_errors(val):
+    """
+    Makes error messages for all deprecation errors
+
+    :param val: ValidateXbrl to check for error
+    :type val: :class:'~arelle.ValidateXbrl.ValidateXbrl
+    :return: No explicit return, though error messages are created
+    :rtype: None
+    """
+    _load_cache(val)
+    for fact in _catch_deprecated_errors(val):
+        val.modelXbrl.error(
+           _CODE_NAME,
+           messages.get_message(_CODE_NAME),
+           concept=fact.concept.label(),
+           modelObject=fact,
+           ruleVersion=_RULE_VERSION
+        )
+
+
+def _deprecated_concept(val, concept):
+    """
+    Returns true if the fact uses a deprecated concept
+
+    :param val: ValidateXbrl to check for deprecated items
+    :type val: :class:'~arelle.ValidateXbrl.ValidateXbrl'
+    :param concept: Concept to check
+    :type concept: :class:'~arelle.ModelFact.Concept'
+    :return: Returns true if the fact uses a deprecated concept
+    :rtype: bool
+    """
+    if concept.name in val.usgaapDeprecations:
+        return True
+    return False
+
+
+def _deprecated_dimension(val, dim_concept):
+    """
+    Returns true if the fact uses a deprecated dimension
+
+    :param val: ValidateXbrl to check for deprecated dimensions
+    :type val: :class:'~arelle.ValidateXbrl.ValidateXbrl'
+    :param dim_concept: Concept to check
+    :type dim_concept: :class:'~arelle.ModelDtsObject.ModelConcept'
+    :return: Returns true if the fact uses a deprecated dimension
+    :rtype: bool
+    """
+    if dim_concept.name in val.usgaapDeprecations:
+        return True
+    return False
+
+
+def _deprecated_member(val, model_dim):
+    """
+    Returns true if the fact uses a deprecated member
+
+    :param val: ValidateXbrl to check for deprecated dimensions
+    :type val: :class:'~arelle.ValidateXbrl.ValidateXbrl'
+    :param modelDim: Concept to check
+    :type modelDim: :class:'~arelle.ModelInstanceObject.ModelDimensionValue'
+    :return: Returns true if the fact uses a deprecated dimension
+    :rtype: bool
+    """
+    if model_dim.isExplicit:
+        member = model_dim.member
+        if member is not None:
+            if member.name in val.usgaapDeprecations:
+                return True
+    return False
+
+
+def _fact_checkable(concept, context):
```

Comment:
You should be able to just pass in a fact and have this function access it's concept and context members, especially considering the name of the function contains fact.